### PR TITLE
Redstone Re-implementation and Async Block PhysicsQueue

### PR DESCRIPTION
**Pull request by PeratX**
Here are some modifications:
- New Redstone System
- Implemented an async block physics and update task
- EffectRange APIs for block physics, reduced unnecessary block updates
- More easier APIs for new Blocks

**Still WIP**
**Have problem with update queue and EffectRange**

### CODE REVIEW

**Comment by NycuRO** [comment]
@PeratX You can create this PR if you added all redstone ,not one files?

**Comment by PeratX** [comment]
I have problem on UpdateQueue and EffectRange.
Waiting for someone to help me.
Without this working properly, the work cannot continue.

**Comment by NycuRO** [comment]
@PeratX I'm created one plugin for add more biomes and more features to Nukkit . Can you help me ?

**Comment by imjack** [comment]
What is the issue?

**Comment by PeratX** [comment]
@imjack See the debug logs. EffectRanges are wrong (THIS_AND_NEIGHBOURS returns only 4). UpdateQueue's sequence seems not right.(See debug logs)

**Comment by PeratX** [comment]
I will go on working on it after EffectRange is fixed. Now I am going to improve the Generator (more populators, implement structures and more!) and implement the ItemMap

**Comment by Snake1999** [comment]
Don't worry, I'll look into it :)

File: src/main/java/cn/nukkit/block/redstone/RedstonePowerMode.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ */
+public enum RedstonePowerMode {
+    /**
+     * Power of all Redstone sources
+     */
+    ALL,
+    /**
+     * Power of all Redstone sources except wire
+     */
+    ALLEXCEPTWIRE
```
Comment (liach) [inline src/main/java/cn/nukkit/block/redstone/RedstonePowerMode.java]: `NONWIRE`?

File: src/main/java/cn/nukkit/level/Level.java [hunk 2]
```diff
@@ -2940,4 +2917,26 @@ public boolean canBlockSeeSky(Vector3 pos) {
         return this.getHighestBlockAt(pos.getFloorX(), pos.getFloorZ()) < pos.getY();
     }
 
+    public boolean runPhysics(){
+        boolean updated = false;
+        updated |= physicsQueue.commitAsyncQueue();
```
Comment (liach) [inline src/main/java/cn/nukkit/level/Level.java]:
```java
boolean updated = physicsQueue.commitAsyncQueue();
```

File: src/main/java/cn/nukkit/utils/Utils.java [hunk 3]
```diff
@@ -126,4 +126,16 @@ public static UUID dataToUUID(byte[]... params) {
         return UUID.nameUUIDFromBytes(stream.toByteArray());
     }
 
+    public static <T extends Collection<?>> T removeDuplicates(T input) {
+        List<Object> unique = new ArrayList<>();
```
Comment (liach) [inline src/main/java/cn/nukkit/utils/Utils.java]: Use hashset?

File: src/main/java/cn/nukkit/utils/Utils.java [hunk 4]
```diff
@@ -126,4 +126,16 @@ public static UUID dataToUUID(byte[]... params) {
...
+        for (Iterator<?> iter = input.iterator(); iter.hasNext(); ) {
+            Object next = iter.next();
+            if (unique.contains(next)) {
+                iter.remove();
+            } else {
+                unique.add(next);
+            }
+        }
+        return input;
+    }
```
Comment (liach) [inline src/main/java/cn/nukkit/utils/Utils.java]: 整个方法也许可以这样写，看看是不是好点：
```java
public static <T extends Collection<?>> T removeDuplicates(T input) {
    if (input instanceof Set) {
        return input;
    }
    Set<Object> unique = new HashSet<>();
    for (Iterator<?> iter = input.iterator(); iter.hasNext(); ) {
        Object next = iter.next();
        if (unique.contains(next)) {
            iter.remove();
        } else {
            unique.add(next);
        }
    }
    return input;
}
```

**Comment by NycuRO** [comment]
@PeratX You can apply changes with @liach ?

File: src/main/java/cn/nukkit/block/redstone/RedstonePowerMode.java (hunk 1 above)
Comment (PeratX) [inline src/main/java/cn/nukkit/block/redstone/RedstonePowerMode.java]: yes

File: src/main/java/cn/nukkit/block/redstone/IndirectRedstoneSource.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * Gets how much redstone power this redstone source block provides to the direction given.<br> This is indirect power, which only powers redstone targets, not other solid blocks
+     *
+     * @param block of this redstone source
+     * @param direction it provides power to
+     * @param powerMode to use to get the power
+     * @return how much power this block provides to the given direction
+     */
+    short getIndirectRedstonePower(Block block, int direction, RedstonePowerMode powerMode);
```
Comment (PeratX) [inline src/main/java/cn/nukkit/block/redstone/IndirectRedstoneSource.java]: yes

**Comment by MagicDroidX** [comment]
@NycuRO I suggest your stopping rushing we nukkit members since we have our brains to think and talk.

**Comment by NycuRO** [comment]
Abandonated?

**Comment by TheModdersDen** [comment]
Is this idea (Redstone) really down the drain in the sense that it is never going to be implemented?

**Comment by Creeperface01** [comment]
@TheModdersDen redstone is already implemented

**Comment by TheModdersDen** [comment]
@Creeperface01 Oh. OK. Is the minecart and or minecart rails implemented yet? That's what I was concerned with specifically...

**Comment by Creeperface01** [comment]
Rails are not implemented, but @kvetinac97 would like to implement it

**Comment by Creeperface01** [comment]
@PeratX how is it going with generator? :D

**Comment by PeratX** [comment]
This branch is deprecated.

**Comment by larryTheCoder** [comment]
I guess you try to say that this branch is outdated xD

**Comment by PeratX** [comment]
No, it's deprecated. I have given up the redstone re-implementation.

**Pull request closed without merging**